src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TrackerObjectManager.NativeAot.cs

```diff
@@ -0,0 +1,543 @@
+            vftbl[3] = (IntPtr)(delegate* unmanaged<IntPtr, IntPtr, int>)&IFindReferenceTargetsCallback_FoundTrackerTarget;
+            return (IntPtr)vftbl;
+        }
+
+        internal static unsafe IntPtr CreateFindReferenceTargetsCallback()
+        {
+            IntPtr wrapperMem = (IntPtr)NativeMemory.Alloc(2 * (nuint)sizeof(IntPtr));
+            *(IntPtr*)(wrapperMem) = CreateDefaultIFindReferenceTargetsCallbackVftbl();
+            *(IntPtr*)(wrapperMem + sizeof(IntPtr)) = wrapperMem;
+            return wrapperMem;
+        }
+    }
+
+    internal readonly struct ComHolder : IDisposable
+    {
+        private readonly IntPtr _ptr;
+
+        internal readonly IntPtr Ptr => _ptr;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ComHolder(IntPtr ptr)
+        {
+            _ptr = ptr;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly void Dispose()
+        {
+            Marshal.Release(_ptr);
+        }
+    }
+
+    // This is used during a GC callback so it needs to be free of any managed allocations.
+    internal unsafe struct DependentHandleList
+    {
+        private int _freeIndex;                // The next available slot
+        private int _capacity;                 // Total numbers of slots available in the list
+        private IntPtr* _pHandles;             // All handles
+        private int _shrinkHint;               // How many times we've consistently seen "hints" that a
+                                               // shrink is needed
+
+        private const int DefaultCapacity = 100;       // Default initial capacity of this list
+        private const int ShrinkHintThreshold = 10;    // The number of hints we've seen before we really
+                                                       // shrink the list
+
+        public bool AddDependentHandle(object target, object dependent)
+        {
+            if (_freeIndex >= _capacity)
+            {
+                // We need a bigger dependent handle array
+                if (!Grow())
+                    return false;
+            }
+
+            IntPtr handle = _pHandles[_freeIndex];
+            if (handle != default)
+            {
+                RuntimeImports.RhHandleSet(handle, target);
+                RuntimeImports.RhHandleSetDependentSecondary(handle, dependent);
+            }
+            else
+            {
+                _pHandles[_freeIndex] = RuntimeImports.RhHandleAllocDependent(target, dependent);
+            }
+
+            _freeIndex++;
+            return true;
+        }
+
+        public bool Reset()
+        {
+            // Allocation for the first time
+            if (_pHandles == null)
+            {
+                _capacity = DefaultCapacity;
+                _pHandles = (IntPtr*)Interop.Ucrtbase.calloc((nuint)_capacity, (nuint)sizeof(IntPtr));
```

Comment:
`calloc` sets the memory to zero. Do you really need the memory zero-initialized here?

Reply:
Yes we do, the logic as ported checks if there was a previously allocated dependency handle at the slot by checking if it is not default / zero and if so overrides the target and dependent for it rather than allocating a new one.